packages/steps/src/BuildStepInput.ts

```diff
@@ -110,20 +112,16 @@ export class BuildStepInput<
     }
 
     const valueDoesNotRequireInterpolation =
-      rawValue === undefined ||
-      typeof rawValue === 'boolean' ||
-      typeof rawValue === 'number' ||
-      typeof rawValue === 'object';
+      rawValue === undefined || typeof rawValue === 'boolean' || typeof rawValue === 'number';
     if (valueDoesNotRequireInterpolation) {
-      const currentTypeName =
-        typeof rawValue === 'object' ? BuildStepInputValueTypeName.JSON : typeof rawValue;
-      if (currentTypeName !== this.allowedValueTypeName && rawValue !== undefined) {
+      if (typeof rawValue !== this.allowedValueTypeName && rawValue !== undefined) {
         throw new BuildStepRuntimeError(
           `Input parameter "${this.id}" for step "${this.stepDisplayName}" must be of type "${this.allowedValueTypeName}".`
         );
       }
       return rawValue as BuildStepInputValueTypeWithRequired<T, R>;
     } else {
+      assert(rawValue !== undefined);
```

Comment:
```suggestion
      // `valueDoesNotRequireInterpolation` checks that `rawValue` is not undefined
      // so this will never be true.
      assert(rawValue !== undefined);
```